packages/statemanager/src/cache/storage.ts

```diff
@@ -0,0 +1,374 @@
+import { bytesToHex, hexStringToBytes } from '@ethereumjs/util'
+import { debug as createDebugLogger } from 'debug'
+import { OrderedMap } from 'js-sdsl'
+
+import { Cache } from './cache'
+import { CacheType } from './types'
+
+import type { CacheOpts } from './types'
+import type { Address } from '@ethereumjs/util'
+import type LRUCache from 'lru-cache'
+
+const LRU = require('lru-cache')
+
+/**
+ * key -> storage mapping
+ *
+ * undefined: storage value is known not to exist in the cache
+ */
+type DiffStorageCacheMap = OrderedMap<string, Uint8Array | undefined>
+type StorageCacheMap = OrderedMap<string, Uint8Array>
+
+export class StorageCache extends Cache {
+  _lruCache: LRUCache<string, StorageCacheMap> | undefined
```

Comment:
As a general question, is there any advantage to using the `LRUCache` for the top level cache if we are using `OrderedMaps` for each individual contract storage cache?  Would it be possible to make the `StorageCacheMap` an LRUCache as well?

Reply:
Yes, this is an intentional decision.

So currently/now the storage cache is implemented in a two way structure. There is the outer account structure, this is the main cache structure which can be switched from LRU to ORDERED_MAP. Every account entry then holds this `StorageCacheMap` where storage slots requested are cached/stored.

This leads to the overall functionality of the cache that - for LRU - cache items are thrown away if the account - as a whole, so any storage slot - hasn't been touched for a certain while. One can argue about that, if this is the most optimal or not, I would say this is at least "optimal enough" and *might* even be the superior solution (with the argument that - if some storage slots from an account are touched *at all*, it is likely that some other slots are also requested).

But these optimal-or-not questions are not the main reason I implemented it this way. I actually had the full other version going at some point in time, this was doing a map (or LRU) from `address_slotKey`-> `slotValue`.

This was then very much throwing (for LRU) just the oldest not-used specific slots away, also structure was somewhat simpler. The unsolvable hurdle I came across along implementation was that we have functionality like `clearTouchedAccounts()`, where it is necessary to re-retrieve all cached slots for one specific account: for this one-layer structure this is just not possible any more (without building up complex side structures).

So far on the context. 😋

For this new structure we only store a couple of storage slots for each account in this additional structure. If we would use an LRU cache for these objects as well that would create a huge overhead, especially on the memory side, if we instantiate these fixed size objects for each and every account, especially regarding that a reasonable cache size is roughly 200.000 or so (that's the client default), so we would create 200.000 LRU objects with this (and we already had memory problems with 1, lol 😋).

Yeah, so. That's the long version of the reasoning togehter with some context.

Reply:
👍 

I guess I chose `OrderedMap` since I assumed it to be faster (at least I think that was the reason, I CAN remember that I HAD a reason and thought about that, lol). We can at some point do experiments, switch. Not totally easy since these iterations would need to be completely rewritten.

Side note: I haven't forgotten all your other review comments in the old PR, just accidentally started by addressing the ones in this newer PR! 🙂